config/config.go

```diff
@@ -0,0 +1,118 @@
+package config
+
+import (
+	"io/ioutil"
+	"path"
+	"gopkg.in/yaml.v2"
+	"github.com/gruntwork-io/boilerplate/util"
+	"fmt"
+	"github.com/gruntwork-io/boilerplate/errors"
+)
+
+const BOILERPLATE_CONFIG_FILE = "boilerplate.yml"
+
+const OPT_TEMPLATE_FOLDER = "template-folder"
+const OPT_OUTPUT_FOLDER = "output-folder"
+const OPT_NON_INTERACTIVE = "non-interactive"
+const OPT_VAR = "var"
+const OPT_VAR_FILE = "var-file"
+
+// The command-line options for the boilerplate app
+type BoilerplateOptions struct {
+	TemplateFolder 	string
+	OutputFolder 	string
+	NonInteractive	bool
+	Vars		map[string]string
+}
+
+// Validate that the options have reasonable values and return an error if they don't
+func (options *BoilerplateOptions) Validate() error {
+	if options.TemplateFolder == "" {
+		return errors.WithStackTrace(TemplateFolderCannotBeEmpty)
+	}
+
+	if !util.PathExists(options.TemplateFolder) {
+		return errors.WithStackTrace(TemplateFolderDoesNotExist(options.TemplateFolder))
+	}
+
+	if options.OutputFolder == "" {
+		return errors.WithStackTrace(OutputFolderCannotBeEmpty)
+	}
+
+	return nil
+}
+
+// The contents of a boilerplate.yml config file
+type BoilerplateConfig struct {
+	Variables []Variable
+}
+
+// A single variable defined in a boilerplate.yml config file
+type Variable struct {
+	Name 	string
+	Prompt 	string
+	Default	string
+}
+
+// Return the default path for a boilerplate.yml config file in the given folder
+func BoilerPlateConfigPath(templateFolder string) string {
+	return path.Join(templateFolder, BOILERPLATE_CONFIG_FILE)
+}
+
+// Load the boilerplate.yml config contents for the folder specified in the given options
+func LoadBoilerPlateConfig(options *BoilerplateOptions) (*BoilerplateConfig, error) {
+	configPath := BoilerPlateConfigPath(options.TemplateFolder)
+
+	if util.PathExists(configPath) {
+		util.Logger.Printf("Loading boilerplate config from %s", configPath)
+		bytes, err := ioutil.ReadFile(configPath)
+		if err != nil {
+			return nil, err
+		}
+
+		return ParseBoilerplateConfig(bytes)
+	} else {
+		util.Logger.Printf("Warning: boilerplate config file not found at %s. No variables will be available while generating.", configPath)
+		return &BoilerplateConfig{}, nil
+	}
+}
+
+// Parse the given configContents as a boilerplate.yml config file
+func ParseBoilerplateConfig(configContents []byte) (*BoilerplateConfig, error) {
+	boilerplateConfig := &BoilerplateConfig{}
+
+	if err := yaml.Unmarshal(configContents, boilerplateConfig); err != nil {
+		return nil, err
+	}
```

Comment:
> What's the advantage of a macro over reflection?

I'm not sure of all of the reasons, but my understanding is that people want to avoid reflection because:

1. It can be slow. I suspect this is one of those performance myths that isn't true anymore in the real world, but if you google "Java reflection slow", you'll get lots and lots of results.
1. It bypasses the type system. If you use reflection, you can do, more or less, whatever you want, and the compiler won't do much about it. That means the serialization/deserialization code could have lots of bugs and the compiler won't help you catch them. I haven't found this to be a problem in real-world serialization code, but perhaps others have, and type system purists avoid reflection whenever they can.
1. Reflection is magical. It's hard to reason about it, it's hard to debug, and it's harder to customize its behavior to do exactly what you want. Again, I'm not sure this is a big problem in real-world serialization code, but functional programming purists tend to avoid reflection whenever they can.

Since Play is all about performance, type safety, and functional programming, it makes sense that they avoid reflection, at least for their Scala API. Their original JSON API had you manually specify how to parse JSON. Later, they removed a lot of that repetitive boilerplate with macros.

To be honest, having used their alternative solution, both with and without macros, it feels like a case of purity over practicality, and I tend to side with the reflection folks on this one. 

> How can you have compile-time driven serialization if the data you receive comes at runtme?

Obviously, the parsing of the data from user requests can only happen at run time. However, with the Play approach, you define a) your data model and b) how to parse arbitrary data into that data model at compile time. This means the compiler can check a lot more of that logic and, in theory, that logic can execute much faster. With the reflection approach, you define a data model at compile time, but the logic of how you convert the runtime data into your data model happens at runtime (via reflection magic) and so the compiler doesn't offer much help, and you may pay a performance penalty. In practice, I haven't seen much of an impact from these differences, especially with thoroughly battle-tested reflection libraries like Jackson, so the compile-time advantages seem mostly theoretical to me.

> Also, isn't use of a macro, which as I understand will generate code at compile-time, more likely to have bugs or edge cases than a native language support?

There is no "native language support". Either you write code that does parsing logic at compile time using macros or at runtime using reflection. Neither is easy to write or debug and both require a fairly deep understanding of the language. Arguably, the macro stuff is the hardest of all to get right, but I suspect the difference in bugginess is not significant.